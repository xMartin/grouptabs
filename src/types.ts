export interface ActionMap {
  createOrUpdate: Entity[];
  delete: Entity[];
}

export enum DocumentType {
  TRANSACTION = "transaction",
  INFO = "info",
}

export enum TransactionType {
  SHARED = "SHARED",
  DIRECT = "DIRECT",
}

export interface Account {
  participant: string;
  amount: number;
}

export interface Transaction {
  id: string;
  type: DocumentType.TRANSACTION;
  tabId: string;
  timestamp: string;
  transactionType: TransactionType;
  description: string;
  date: string;
  participants: Account[];
}

export interface Info {
  id: string;
  type: DocumentType.INFO;
  tabId: string;
  name: string;
}

export type Entity = Transaction | Info;

export interface Tab {
  id: string;
  info: Info;
  mostRecentTransaction?: Transaction;
}

export enum TransactionFormParticipantStatus {
  NONE = 0,
  JOINED = 1,
  PAID = 2,
}

export enum TransactionFormParticipantInputType {
  NEW,
  EXISTING,
}

export interface TransactionFormSharedState {
  id: string;
  inputType: TransactionFormParticipantInputType;
  participant?: string;
  status: TransactionFormParticipantStatus;
  amount?: number;
}

interface TransactionFormDirectState {
  from?: string;
  fromNew?: string;
  to?: string;
  toNew?: string;
  amount?: number;
  options: string[];
}

export interface TransactionFormState {
  transactionType: TransactionType;
  date: string;
  description?: string;
  shared: TransactionFormSharedState[];
  direct: TransactionFormDirectState;
}
