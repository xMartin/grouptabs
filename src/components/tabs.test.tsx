import renderer from "react-test-renderer";
import Tabs from "./tabs";
import { DocumentType } from "../types";

it("renders tab view", () => {
  const tree = renderer
    .create(
      <Tabs
        data={[
          {
            id: "123",
            info: {
              id: "123",
              type: DocumentType.INFO,
              tabId: "567",
              name: "TAB NAME",
            },
          },
        ]}
        onTabClick={vi.fn()}
        onCreateTabInputChange={vi.fn()}
        onCreateNewTab={vi.fn()}
        onImportTabInputChange={vi.fn()}
        onImportTab={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders empty tab view", () => {
  const tree = renderer
    .create(
      <Tabs
        data={[]}
        onTabClick={vi.fn()}
        onCreateTabInputChange={vi.fn()}
        onCreateNewTab={vi.fn()}
        onImportTabInputChange={vi.fn()}
        onImportTab={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
