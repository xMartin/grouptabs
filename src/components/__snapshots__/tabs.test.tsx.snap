// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty tab view 1`] = `
<div
  className="scene tabsScene"
>
  <div
    className="header header-brand"
  >
    <div
      className="brand"
    >
      <img
        alt=""
        className="logo"
        src="logo.png"
      />
      <svg
        height="5.954884mm"
        viewBox="0 0 30.910572 5.954884"
        width="30.910572mm"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fillOpacity=".969337"
          strokeWidth="3.92"
          transform="translate(7.760592 -61.592628)"
        >
          <path
            d="m-6.425955 67.547513q-.7337773 0-1.0724437-.564444-.3386664-.578556-.2398887-1.396999l.4092219-3.471331q.014111-.141111.084667-.296333.070556-.155222.3527775-.155222.282222 0 .3386664.155222.056444.155222.028222.338666l-.3668887 3.428998q-.056444.507999.070556.832555.1269999.310444.4374442.310444.3527775 0 .705555-.282222.3668886-.282222.7902216-.945444.423333-.677333.9454437-1.820332l-.1552221 1.834443q-.5221107.903111-1.0865547 1.467555-.5503329.564444-1.2417768.564444zm2.1448872-.01411q-.2116665 0-.282222-.112888-.056444-.127-.042333-.338667l.5362218-4.967107q.014111-.141111.098778-.296333.084667-.155222.3527775-.155222.2681109 0 .3386664.155222.070555.155222.042333.338666l-.5221107 4.924774q-.028222.225778-.1552221.338667-.1128888.112888-.3668886.112888z"
          />
          <path
            d="m-1.8680943 67.533401q-.2681109 0-.3386664-.155222-.070556-.169333-.056444-.366888l.564444-4.896552q.028222-.225778.1552221-.338666.1269999-.112889.3809997-.112889.21166649 0 .26811089.127.0564444.112889.0282222.324555l-.56444399 4.896552q-.028222.169333-.070556.352777-.042333.169333-.3668886.169333zm3.245553 0q-.2681109 0-.3245553-.155222-.0564444-.169333-.042333-.366888l.4092216-3.400775q.070556-.550333-.084667-.860777-.141111-.324556-.43744409-.324556-.52211071 0-1.07244361.733777-.53622181.733778-1.3264434 2.469443l.1693332-1.876776q.47977739-1.001888 1.03011029-1.580444.55033291-.578555 1.24177682-.578555.71966609 0 1.07244359.550333.3527775.536222.2398887 1.467555l-.423333 3.400775q-.014111.169333-.070556.352777-.056444.169333-.3809997.169333z"
          />
          <path
            d="m5.3708949 67.547513q-.5503329 0-.9454437-.254-.3809997-.254-.6208884-.677333-.2257776-.423333-.3104442-.945444-.084667-.536222-.028222-1.086554.1128888-.973666.4656663-1.636888.3668886-.663222.9031104-1.001888.5362218-.352778 1.1571102-.352778.4092219 0 .7619994.169334.3668886.155222.5926663.423333.2257776.253999.2257776.564444 0 .324555-.3245553.719666-.3104443.380999-.8184439.77611-.5079996.395111-1.0865547.776111-.564444.366888-1.0724436.691444l-.098778-.818444q.3809997-.239889.8184438-.536222.4515552-.296333.846666-.592666.3951108-.310444.6349995-.550333.2539998-.254.2539998-.395111 0-.155222-.1834443-.296333-.1834443-.155222-.5785551-.155222-.6067773 0-1.0724436.592666-.4656663.592666-.5785551 1.66511-.056444.550333.028222 1.044221.098778.493889.3668886.804333.282222.310444.7478883.310444.3245553 0 .6491106-.239889.3245553-.253999.5503329-.705555.070556-.126999.2257776-.169333.1693332-.04233.3245553.05644.1552222.08467.1975555.239888.042333.155222-.028222.282222-.3104443.564444-.8043328.931333-.4938885.366889-1.1994435.366889z"
          />
          <path
            d="m9.6606577 67.180624-.9736659-4.811885q-.056444-.338667.056444-.522111.1128888-.183444.3386664-.183444.1269999 0 .2398887.09878.1128888.08467.1552221.26811l.7761106 4.317997-.141111-.04233 1.890887-4.275664q.112889-.211666.225778-.282222.112889-.08467.239889-.08467.155222 0 .253999.08467.112889.08467.141111.239889.02822.155222-.07056.366889l-2.243665 4.854218q-.09878.197556-.211666.268111-.112889.05644-.268111.05644-.1411114 0-.2681113-.07055-.1128888-.08467-.141111-.282222z"
          />
          <path
            d="m15.333296 67.547513q-.550333 0-.945443-.254-.381-.254-.620889-.677333-.225777-.423333-.310444-.945444-.08467-.536222-.02822-1.086554.112889-.973666.465666-1.636888.366889-.663222.903111-1.001888.536221-.352778 1.15711-.352778.409222 0 .761999.169334.366889.155222.592666.423333.225778.253999.225778.564444 0 .324555-.324555.719666-.310445.380999-.818444.77611-.508.395111-1.086555.776111-.564444.366888-1.072444.691444l-.09878-.818444q.381-.239889.818444-.536222.451555-.296333.846666-.592666.39511-.310444.634999-.550333.254-.254.254-.395111 0-.155222-.183444-.296333-.183445-.155222-.578555-.155222-.606778 0-1.072444.592666-.465666.592666-.578555 1.66511-.05645.550333.02822 1.044221.09878.493889.366889.804333.282222.310444.747888.310444.324555 0 .649111-.239889.324555-.253999.550332-.705555.07056-.126999.225778-.169333.169333-.04233.324555.05644.155222.08467.197556.239888.04233.155222-.02822.282222-.310445.564444-.804333.931333-.493889.366889-1.199444.366889z"
          />
          <path
            d="m19.002172 67.533401q-.268111 0-.338667-.155222-.07056-.169333-.05644-.366888l.564444-4.896552q.02822-.225778.155222-.338666.127-.112889.381-.112889.211666 0 .268111.127.05644.112889.02822.324555l-.564444 4.896552q-.02822.169333-.07056.352777-.04233.169333-.366888.169333zm3.245553 0q-.268111 0-.324556-.155222-.05644-.169333-.04233-.366888l.409222-3.400775q.07056-.550333-.08467-.860777-.141111-.324556-.437444-.324556-.52211 0-1.072443.733777-.536222.733778-1.326444 2.469443l.169333-1.876776q.479778-1.001888 1.030111-1.580444.550333-.578555 1.241777-.578555.719666 0 1.072443.550333.352778.536222.239889 1.467555l-.423333 3.400775q-.01411.169333-.07056.352777-.05644.169333-.380999.169333z"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="content content-with-footer"
  >
    <main>
      <div
        className="empty-info"
      >
        <p>
          Track shared expenses in a group of people. You can create groups for projects or topics like "Summer roadtrip" or "Badminton".
        </p>
        <p>
          Start by creating your first group:
        </p>
      </div>
      <div
        className="form-row"
      >
        <form
          className="create-form"
          onSubmit={[Function]}
        >
          <input
            className="full-width"
            onChange={[Function]}
            placeholder="Group title …"
            type="text"
            value=""
          />
          <button
            className="create"
          >
            Create
          </button>
        </form>
      </div>
      <div
        className="form-row"
      >
        <p
          className="fake-link"
          onClick={[Function]}
        >
          Import group
        </p>
      </div>
    </main>
    <footer>
      Version: 
      N/A
       –
       
      <a
        href="https://getuneven.com/"
        rel="noopener"
        target="_blank"
      >
        getuneven.com
      </a>
    </footer>
  </div>
</div>
`;

exports[`renders tab view 1`] = `
<div
  className="scene tabsScene"
>
  <div
    className="header header-brand"
  >
    <div
      className="brand"
    >
      <img
        alt=""
        className="logo"
        src="logo.png"
      />
      <svg
        height="5.954884mm"
        viewBox="0 0 30.910572 5.954884"
        width="30.910572mm"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fillOpacity=".969337"
          strokeWidth="3.92"
          transform="translate(7.760592 -61.592628)"
        >
          <path
            d="m-6.425955 67.547513q-.7337773 0-1.0724437-.564444-.3386664-.578556-.2398887-1.396999l.4092219-3.471331q.014111-.141111.084667-.296333.070556-.155222.3527775-.155222.282222 0 .3386664.155222.056444.155222.028222.338666l-.3668887 3.428998q-.056444.507999.070556.832555.1269999.310444.4374442.310444.3527775 0 .705555-.282222.3668886-.282222.7902216-.945444.423333-.677333.9454437-1.820332l-.1552221 1.834443q-.5221107.903111-1.0865547 1.467555-.5503329.564444-1.2417768.564444zm2.1448872-.01411q-.2116665 0-.282222-.112888-.056444-.127-.042333-.338667l.5362218-4.967107q.014111-.141111.098778-.296333.084667-.155222.3527775-.155222.2681109 0 .3386664.155222.070555.155222.042333.338666l-.5221107 4.924774q-.028222.225778-.1552221.338667-.1128888.112888-.3668886.112888z"
          />
          <path
            d="m-1.8680943 67.533401q-.2681109 0-.3386664-.155222-.070556-.169333-.056444-.366888l.564444-4.896552q.028222-.225778.1552221-.338666.1269999-.112889.3809997-.112889.21166649 0 .26811089.127.0564444.112889.0282222.324555l-.56444399 4.896552q-.028222.169333-.070556.352777-.042333.169333-.3668886.169333zm3.245553 0q-.2681109 0-.3245553-.155222-.0564444-.169333-.042333-.366888l.4092216-3.400775q.070556-.550333-.084667-.860777-.141111-.324556-.43744409-.324556-.52211071 0-1.07244361.733777-.53622181.733778-1.3264434 2.469443l.1693332-1.876776q.47977739-1.001888 1.03011029-1.580444.55033291-.578555 1.24177682-.578555.71966609 0 1.07244359.550333.3527775.536222.2398887 1.467555l-.423333 3.400775q-.014111.169333-.070556.352777-.056444.169333-.3809997.169333z"
          />
          <path
            d="m5.3708949 67.547513q-.5503329 0-.9454437-.254-.3809997-.254-.6208884-.677333-.2257776-.423333-.3104442-.945444-.084667-.536222-.028222-1.086554.1128888-.973666.4656663-1.636888.3668886-.663222.9031104-1.001888.5362218-.352778 1.1571102-.352778.4092219 0 .7619994.169334.3668886.155222.5926663.423333.2257776.253999.2257776.564444 0 .324555-.3245553.719666-.3104443.380999-.8184439.77611-.5079996.395111-1.0865547.776111-.564444.366888-1.0724436.691444l-.098778-.818444q.3809997-.239889.8184438-.536222.4515552-.296333.846666-.592666.3951108-.310444.6349995-.550333.2539998-.254.2539998-.395111 0-.155222-.1834443-.296333-.1834443-.155222-.5785551-.155222-.6067773 0-1.0724436.592666-.4656663.592666-.5785551 1.66511-.056444.550333.028222 1.044221.098778.493889.3668886.804333.282222.310444.7478883.310444.3245553 0 .6491106-.239889.3245553-.253999.5503329-.705555.070556-.126999.2257776-.169333.1693332-.04233.3245553.05644.1552222.08467.1975555.239888.042333.155222-.028222.282222-.3104443.564444-.8043328.931333-.4938885.366889-1.1994435.366889z"
          />
          <path
            d="m9.6606577 67.180624-.9736659-4.811885q-.056444-.338667.056444-.522111.1128888-.183444.3386664-.183444.1269999 0 .2398887.09878.1128888.08467.1552221.26811l.7761106 4.317997-.141111-.04233 1.890887-4.275664q.112889-.211666.225778-.282222.112889-.08467.239889-.08467.155222 0 .253999.08467.112889.08467.141111.239889.02822.155222-.07056.366889l-2.243665 4.854218q-.09878.197556-.211666.268111-.112889.05644-.268111.05644-.1411114 0-.2681113-.07055-.1128888-.08467-.141111-.282222z"
          />
          <path
            d="m15.333296 67.547513q-.550333 0-.945443-.254-.381-.254-.620889-.677333-.225777-.423333-.310444-.945444-.08467-.536222-.02822-1.086554.112889-.973666.465666-1.636888.366889-.663222.903111-1.001888.536221-.352778 1.15711-.352778.409222 0 .761999.169334.366889.155222.592666.423333.225778.253999.225778.564444 0 .324555-.324555.719666-.310445.380999-.818444.77611-.508.395111-1.086555.776111-.564444.366888-1.072444.691444l-.09878-.818444q.381-.239889.818444-.536222.451555-.296333.846666-.592666.39511-.310444.634999-.550333.254-.254.254-.395111 0-.155222-.183444-.296333-.183445-.155222-.578555-.155222-.606778 0-1.072444.592666-.465666.592666-.578555 1.66511-.05645.550333.02822 1.044221.09878.493889.366889.804333.282222.310444.747888.310444.324555 0 .649111-.239889.324555-.253999.550332-.705555.07056-.126999.225778-.169333.169333-.04233.324555.05644.155222.08467.197556.239888.04233.155222-.02822.282222-.310445.564444-.804333.931333-.493889.366889-1.199444.366889z"
          />
          <path
            d="m19.002172 67.533401q-.268111 0-.338667-.155222-.07056-.169333-.05644-.366888l.564444-4.896552q.02822-.225778.155222-.338666.127-.112889.381-.112889.211666 0 .268111.127.05644.112889.02822.324555l-.564444 4.896552q-.02822.169333-.07056.352777-.04233.169333-.366888.169333zm3.245553 0q-.268111 0-.324556-.155222-.05644-.169333-.04233-.366888l.409222-3.400775q.07056-.550333-.08467-.860777-.141111-.324556-.437444-.324556-.52211 0-1.072443.733777-.536222.733778-1.326444 2.469443l.169333-1.876776q.479778-1.001888 1.030111-1.580444.550333-.578555 1.241777-.578555.719666 0 1.072443.550333.352778.536222.239889 1.467555l-.423333 3.400775q-.01411.169333-.07056.352777-.05644.169333-.380999.169333z"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="content content-with-footer"
  >
    <main>
      <div
        className="row tabs"
      >
        <button
          className="full-width list-item-with-arrow"
          onClick={[Function]}
        >
          TAB NAME
        </button>
      </div>
      <div
        className="form-row"
      >
        <form
          className="create-form"
          onSubmit={[Function]}
        >
          <input
            className="full-width"
            onChange={[Function]}
            placeholder="Group title …"
            type="text"
            value=""
          />
          <button
            className="create"
          >
            Create
          </button>
        </form>
      </div>
      <div
        className="form-row"
      >
        <p
          className="fake-link"
          onClick={[Function]}
        >
          Import group
        </p>
      </div>
    </main>
    <footer>
      Version: 
      N/A
       –
       
      <a
        href="https://getuneven.com/"
        rel="noopener"
        target="_blank"
      >
        getuneven.com
      </a>
    </footer>
  </div>
</div>
`;
